### PR TITLE
Progress plots, tutorial 5, saving and loading pipelines

### PR DESCRIPTION
**Pull request by luizfelippesr**
Includes some standardized progress reports which allow one to better keep track of how the samplers are doing.

**To do (progress reports:)**
- [x] Finish PR #57 
- [x] Progress reports for MultiNest
- ~Progress reports for Dynesty~
- ~Progress reports for UltraNest~

**To do (tutorial_five)**
- [x] Check text
- [x] Integrate to docs

**To do (saving and loading)**
- [x] Basic functionality
- [x] Working directory adjustments
- [x] Docs and/or tutorial (_update: tutorial only, docs will/can come later)_

### CODE REVIEW

**Comment by luizfelippesr** [comment]
UltraNest and Dynesty have their own ways of reporting progress and write diagnostics. So, perhaps we shouldn't be using our time right now trying to write a new way to track the progress for them.  Later, the present simple progress tracker can be substituted by something more general and/or polished.

**Comment by luizfelippesr** [comment]
When reviewing this PR, please ignore /examples/example_pipeline.py, which is being looked at in PR #60

File: imagine/priors/basic_priors.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ class GaussianPrior(ScipyPrior):
     def __init__(self, mu=0.0, sigma=1.0, xmin=None, xmax=None, unit=None,
```
Comment (shutsch) [inline imagine/priors/basic_priors.py:77]: Also here I think we should get rid of defaults

File: runs/tutorial_example/chains/multinest_.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+    0.322344980648892928E-26    0.153469400698917320E+04    0.301938202186512505E+01    0.275005312893483840E+02
```
Comment (shutsch) [inline runs/tutorial_example/chains/multinest_.txt:1]: Presumably all these files are for the Jupyter notebooks?

File: imagine/pipelines/pipeline.py [hunk 3]
```diff
@@ -66,18 +71,30 @@ class Pipeline(BaseClass, metaclass=abc.ABCMeta):
...
     chains_directory : str
-        Path of the directory where the chains should be saved
+        Path of the directory where the chains should be saved. By default,
+        this is saved to a 'chains' subdirectory of `run_directory`.
     prior_correlations : dict
         Dictionary used to set up prior distribution correlations. If two
         parameters are A and B are correlated a priori, an entry should be
         added to the prior_correlations dictionary in the form
         `(name_A, name_B): True`, to extract the correlation from the samples
         (in the case of CustomPriors) or `(name_A, name_B): value` otherwise.
```
Comment (shutsch) [inline imagine/pipelines/pipeline.py:84]: I think we should explain more what we mean with 'value' here (i.e the nonlinear correlation coefficients). 
I will have a look the next days.

**Review by shutsch — APPROVED** [review]
I agree with all changes and I have tested the tutorials and a few other things.
I quite like the possibility of saving the pipeline and restarting it so easily. I also like the progress plots. 
There are a few comments in my review, but I think none of them directly concerns this pull request.
This is mostly stuff that came to my eyes while reading through the changes, which would I have otherwise forgotten.

File: imagine/pipelines/multinest_pipeline.py [hunk 4]
```diff
@@ -102,7 +104,7 @@ def call(self, **kwargs):
         # Resets internal state
         self.tidy_up()
 
-        default_solve_params = {'resume': False,
+        default_solve_params = {'resume': True,
                                 'n_live_points': 400,
```
[the remark this answers is not in the document]
Comment (1313e) [reply imagine/pipelines/multinest_pipeline.py:108]: Pipelines should work as much out-of-the-box as possible.
Setting somewhat realistic and viable default parameters is always a good idea, as the user can adapt those for their own use.

File: doc/source/index.rst [hunk 5]
```diff
@@ -57,6 +57,7 @@ please refer to the
    tutorial_hammurabi
    tutorial_priors
    tutorial_masks
+   tutorial_five
```
[the remark this answers is not in the document]
Comment (1313e) [reply doc/source/index.rst:60]: `tutorial_wmap` could work just fine.

**Review by 1313e — APPROVED** [review]
I am fine with this being merged as well.
Remember btw that CMasher is also on conda-forge, and can thus be added to the `imagine_conda_env.yml` file.

File: examples/basic_pipeline.py [hunk 6]
```diff
@@ -170,8 +145,10 @@ def basic_pipeline_run(pipeline_class=img.pipelines.MultinestPipeline,
     # Prepares the random magnetic field factory
     B_factory = testFields.NaiveGaussianMagneticFieldFactory(grid=one_d_grid)
     B_factory.active_parameters = ('a0','b0')
-    B_factory.priors ={'a0': img.priors.FlatPrior(interval=[-5, 5]*u.microgauss),
-                      'b0': img.priors.FlatPrior(interval=[2, 10]*u.microgauss)}
+    B_factory.priors ={'a0': img.priors.FlatPrior(xmin=-5*u.microgauss,
+                                                  xmax=5*u.microgauss),
```
Comment (1313e) [inline examples/basic_pipeline.py:149]: Why the `xmin` and `xmax` here?
Don't explicitly set them, as it is incredibly obvious for the user that that is what the two arguments are that the class takes.

File: imagine/priors/basic_priors.py (hunk 1 above)
Comment (1313e) [reply imagine/priors/basic_priors.py:77]: I would also change the order to make `xmin` and `xmax` come first, as it is with other priors.
Therefore, extra information is given after the usual information.

**Comment by 1313e** [comment]
One thing though: There are many auxiliary files in there that should not be tracked by Git.
For example, as @shutsch pointed out, the files in the `chains` folder.
Why are these currently being tracked by Git?

File: examples/basic_pipeline.py [hunk 7]
```diff
@@ -218,50 +196,63 @@ def basic_pipeline_run(pipeline_class=img.pipelines.MultinestPipeline,
...
     # Sets up the sampler to be used
     pipeline_class = img.pipelines.UltranestPipeline
     # Set some controller parameters that are specific to UltraNest.
-    sampling_controllers = {'dlogz': 0.05,
-                            'min_num_live_points': 400,
-                            'min_ess': 1000}
+    sampling_controllers = {'dlogz': 0.5,
+                            'dKL':0.1,
+                            'min_num_live_points': 500,
+                            'min_ess': 1000,
```
[the remark this answers is not in the document]
Comment (luizfelippesr) [reply examples/basic_pipeline.py:209]: I have kept the native argument names used by each sampler. The main reason was a practical one: there are too many of them, and the effort to "translate" and document them becomes simply too big.

File: doc/source/index.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (luizfelippesr) [reply doc/source/index.rst:60]: I will change to tutorial_wmap, then.

File: imagine/pipelines/multinest_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (luizfelippesr) [reply imagine/pipelines/multinest_pipeline.py:108]: I agree with Ellert with the out-of-the-box idea. The only condition is that the user should be able to discover, in the end, which parameters they actually used (which is the case, by inspecting pipeline.sampling_controllers). This allows someone clueless to run, find a result (or a problem) and find out why on managed to get that result (or that problem).

File: imagine/priors/basic_priors.py (hunk 1 above)
Comment (luizfelippesr) [reply imagine/priors/basic_priors.py:77]: Indeed, defaults here are bad. I change the order as well.

File: runs/tutorial_example/chains/multinest_.txt (hunk 2 above)
Comment (luizfelippesr) [reply runs/tutorial_example/chains/multinest_.txt:1]: Yes, these are example chains, to allow tutorial_five (tutorial_wmap) to show something interesting sooner.

File: examples/basic_pipeline.py (hunk 6 above)
Comment (luizfelippesr) [reply examples/basic_pipeline.py:149]: I'll amend this.

**Comment by luizfelippesr** [comment]
> One thing though: There are many auxiliary files in there that should not be tracked by Git.
> For example, as @shutsch pointed out, the files in the `chains` folder.
> Why are these currently being tracked by Git?

These are example chains, to allow tutorial_five (tutorial_wmap) to show something interesting sooner.  In the future, we may consider tweaking the save/load pipeline functions keep them in the pipeline state file (or in a separate chain file).

File: imagine/priors/basic_priors.py (hunk 1 above)
Comment (luizfelippesr) [reply imagine/priors/basic_priors.py:77]: On the other hand, now it is possible to set up a `GaussianPrior` without specifying `xmin` and `xmax`... so, I think I will keep the order for now

File: examples/basic_pipeline.py (hunk 6 above)
Comment (luizfelippesr) [reply examples/basic_pipeline.py:149]: Actually, I will leave this for #60

**Pull request merged**